feat: Add telemetry support.

ehsavoie opened the pull request:
Adding support for OpenTelemetry

Fixes #388 🦕

brunobat commented on server-common/src/main/java/io/a2a/server/interceptors/AttributeExtractor.java:
```diff
@@ -0,0 +1,8 @@
+import java.util.Map;
+
+@FunctionalInterface
+public interface AttributeExtractor {
```
There is also https://opentelemetry.io/docs/zero-code/java/agent/annotations/#adding-attributes-to-the-span-with-spanattribute

brunobat commented on server-common/src/main/java/io/a2a/server/interceptors/Kind.java:
```diff
@@ -0,0 +1,5 @@
+package io.a2a.server.interceptors;
+
+public enum Kind {
```
You are duplicating the OpenTelemetry API...

brunobat reviewed (commented):
I think you need to decide if you are creating your own instrumentation annotations and API or if you just reuse the ones provided by OTel.

Also, the attribute names, should follow these semantic conventions, as much as possible: https://opentelemetry.io/docs/specs/semconv/gen-ai/gen-ai-agent-spans/

brunobat commented on examples/helloworld/client/src/main/java/io/a2a/examples/helloworld/HelloWorldClient.java:
```diff
@@ -108,4 +115,16 @@ public static void main(String[] args) {
-} 
\ No newline at end of file
+    static OpenTelemetry initOpenTelemetry() {
+        SdkTracerProvider sdkTracerProvider = SdkTracerProvider.builder()
```
You can have a new module that creates this... I mention this later.

brunobat replied:
this is an example. that's fine

kabir commented on client/base/src/main/java/io/a2a/client/ClientBuilder.java:
```diff
@@ -37,7 +43,8 @@ public class ClientBuilder {
     private final List<BiConsumer<ClientEvent, AgentCard>> consumers = new ArrayList<>();
-    private @Nullable Consumer<Throwable> streamErrorHandler;
+    private @Nullable
+    Consumer<Throwable> streamErrorHandler;
```
This looks a bit strange :-)

The original looks better, but I think 

````
@Nullable
private Consumer<Throwable> streamErrorHandler;
````
would look better. But I see we have `private @Nullable ....` 'everywhere' :-)

kabir commented on transport/jsonrpc/src/main/java/io/a2a/transport/jsonrpc/handler/JsonRPCAttributeExtractor.java:
```diff
@@ -0,0 +1,41 @@
+                    return Map.of("body", parameters[0].toString(), "extensions", context.getActivatedExtensions().stream().collect(Collectors.joining(",")), "a2a.method", (String) context.getState().get(METHOD_NAME_KEY));
+                }
+                default -> {
+                    return Collections.emptyMap();
```
warn/error

kabir reviewed (commented):
A few things which may or may not be needed :-)

kabir reviewed (dismissed):
Actually, @ehsavoie it looks like you are adding Otel tracing to the example. But I don't know how to enable this etc.
Can you update the README with instructions?

Maybe we could also make this example testable in the CI, like we do for the cloud-deployment one? 

Also, it would be good to have some tests in the extras module

kabir commented on client/base/src/main/java/io/a2a/client/ClientBuilder.java:
```diff
@@ -108,7 +115,8 @@ public class ClientBuilder {
     private final List<BiConsumer<ClientEvent, AgentCard>> consumers = new ArrayList<>();
-    private @Nullable Consumer<Throwable> streamErrorHandler;
+    private @Nullable
+    Consumer<Throwable> streamErrorHandler;
```
This still looks weird :-)

kabir reviewed (commented):
@ehsavoie 2 small things.

If the JBang validator is tricky, open an issue and link it here

ehsavoie commented on extras/opentelemetry/src/main/java/io/a2a/extras/opentelemetry/OpenTelemetryClientTransport.java:
```diff
@@ -0,0 +1,322 @@
+import java.util.stream.Collectors;
+import org.jspecify.annotations.Nullable;
+
+public class OpenTelemetryClientTransport implements ClientTransport {
```
I think I will also need to avoid tracing the full request / response

brunobat commented on extras/opentelemetry/integration-tests/src/test/java/io/a2a/extras/opentelemetry/it/OpenTelemetryIT.java:
```diff
@@ -0,0 +1,325 @@
+ * OpenTelemetry spans are created correctly for A2A operations.
+ */
+@QuarkusIntegrationTest
+class OpenTelemetryIT extends OpenTelemetryTest {
```
Test classes annotated with `@QuarkusIntegrationTest` will only run in native mode. 
Now you don't have any tests running on native. Usually you have a test class running with `@QuarkusTest` and the another one called *IT annotated with `@QuarkusIntegrationTest` that will run Quarkus native mode tests with graalVM

ehsavoie committed:
feat: Add OpenTelemetry tracing support

Adds OpenTelemetry integration for distributed tracing:

- Client: OpenTelemetryClientTransport decorator wrapping all client operations
  with span creation and context propagation via ClientTransportWrapper SPI
- Server: OpenTelemetryRequestHandlerDecorator CDI decorator for request handlers
- Common: A2AObservabilityNames for standard observability attributes
- Integration tests with comprehensive coverage for client and server tracing
- Updated HelloWorld examples with OTEL configuration and documentation

Fixes #388

Signed-off-by: Emmanuel Hugonnet <ehugonne@redhat.com>
